Feature 449943 Safeguard Custom Platform Script for Wikipedia.com

PrabhaQuest opened the pull request:
Adding the custom platform script for Wikipedia.

PrabhaQuest committed:
Added the custom platform script for Wikipedia.

Kevin-Andrew commented on SampleScripts/HTTP/CustomWikipedia.json:
```diff
@@ -0,0 +1,228 @@
+      { "HttpProxyPassword": { "Type": "Secret", "Description": "HTTP Proxy Password", "Required": false } }
+    ],
+    "Do": [
+      { "Status": { "Type": "Changing", "Percent": 20, "Message": { "Name": "ChangingPassword", "Parameters": [ "%AccountUserName%" ] } } },
```
Let's change this `Percent` value to `10`, like it is in the `CheckPassword` operation. Line 57 below has a `Status` with a `Percent` of `20`. So I don't think it makes much sense to have them both the same.

PrabhaQuest replied:
Yes as you said the Percentage was not proper, I have changed it to 10 so that it makes sense.

Kevin-Andrew commented on SampleScripts/HTTP/CustomWikipedia.json:
```diff
@@ -0,0 +1,228 @@
+          "If": "AccountPassword.Equals(NewPassword)", "Then": 
+          [
+            { "Status": { "Type": "Changing", "Percent": 80, "Message": { "Name": "CurrentAndNewPasswordsAreIdentical", "Parameters": [ "%AccountUserName%" ] } } },
+            { "Log": { "Text": "Current and New Password are Identical" } },
```
Change the case of some of the words to read as a regular sentence. "The current and new password are identical."

PrabhaQuest replied:
I have changed it to regular sentence.

Kevin-Andrew commented on SampleScripts/HTTP/CustomWikipedia.json:
```diff
@@ -0,0 +1,228 @@
+          }
+        },
+        { "ExtractJsonObject": { "JsonObjectName": "LoginPostResponse", "Name": "LoginResultJson" } },
+        { "Condition": {
+            "If": "!LoginResultJson.login.result.Value.Equals(\"Success\")", "Then": 
+            [
+              { "Log": { "Text": "Unsuccessful Login" } },
+              { "Return": { "Value": false } }
+            ]
+          }
+        },
+        { "Condition": {
+            "If": "LoginResultJson.login.result.Value.Equals(\"Success\")", "Then": 
+            [
+              { "Log": { "Text": "Authentication successful" } },
+              { "Return": { "Value": true } }
+            ]
+          }
+        },
```
Is the response automatically logged? If not, we should consider logging the response for a failed login. That way, a person can look at the log and perhaps get an idea of why it failed. Right now, just manually logging "Unsuccessful login" probably isn't too helpful.

And then, we can actually get rid of the first `Condition`. Instead, this can be rewritten as just the second `Condition` block. Meaning, all you do is check for the `Success`, and if `true`, the function will return `true` and exit.

But then, if it wasn't successful, there is no need for any other checks. We can simply log a failure message, and include the response, then set the `Status` as is being done, and return `false` as is being done.

PrabhaQuest replied:
I have logged the reson for failure when it fails. Though always the response is logged for all the request, I log the specific reason alone from the response during the failure.

Kevin-Andrew commented on SampleScripts/HTTP/CustomWikipedia.json:
```diff
@@ -0,0 +1,228 @@
+             "IsSecret":  true
+           }
+         },
+         { "SetItem": { "Name": "RequestBody", "Value": "changeauthtoken=%{EncodedCsrfToken}%&password=%EncodedNewPassword%&retype=%EncodedNewPassword%" , "IsSecret": true} },
```
You use the expression substitution syntax, `%{ }%` on the `EncodedCsrfToken`, but on the other two variables, you just use the regular simple string substitution, which is fine and is all that is needed, correct? So for consistency, we can make the `EncodedCsrfToken` the same and just use `%EncodedCsrfToken%`.

PrabhaQuest replied:
Yes, both the ways it works. I have removed the '{' and '}' to make it consistent in the script.

Kevin-Andrew commented on SampleScripts/HTTP/CustomWikipedia.json:
```diff
@@ -0,0 +1,228 @@
+            "ResultVariable": "EncodedUserName"
+          }
+        },
+        { "SetItem": { "Name": "RequestBody", "Value": "lgpassword=%{EncodedPassword}%&lgtoken=%{EncodedLoginToken}%" , "IsSecret": true } },
```
See the other comment about the expression substitution syntax, `%{ }%`, versus simple string substitution. All we need is the simple string substitution, correct? Either way works, but I just want it to be consistent.

PrabhaQuest replied:
I have removed the '{' and '}' to make it consistent in the script.

PrabhaQuest committed:
Feature 449943 Safeguard Custom Platform Script for Wikipedia.com

This script is for including Wikipedia as a new custom platform in SPP.

Kevin-Andrew commented on SampleScripts/HTTP/CustomWikipedia.json:
```diff
@@ -134,7 +127,8 @@
             [
-              { "Log": { "Text": "Unsuccessful Login" } },
+              { "Log": { "Text": "%{ LoginResultJson.login.reason.Value }%" } },
+              { "Status": { "Type": "Checking", "Percent": 80, "Message": { "Name": "LoggingInWithAccountFailed", "Parameters": [ "%AccountUserName%" ] } } },
```
The `LoggingInWithAccountFailed` message takes two parameters. See https://github.com/Kevin-Andrew/Hercules/blob/master/Source/Rsms.Public/StatusOutput/PlatformStatusStrings.resx#L465

PrabhaQuest replied:
Yes I have missed that. I have reverified and corrected those.

Kevin-Andrew commented on SampleScripts/HTTP/CustomWikipedia.json:
```diff
@@ -216,8 +210,7 @@
         { "Condition": {
             "If": "PasswordResponseJson.ToString().Contains(\"error\")", "Then": 
             [
```
No need for this `Condition` block. We already know that the request wasn't successful. So why would we try to check for a specific "error"? Just log the entire `PasswordResponseJson.ToString()` response, and return `false` from the method.

PrabhaQuest replied:
I have removed the Condition block. I have done this change for Login as well as Password change. If there is a failure I log the response and return False.

PrabhaQuest committed:
Feature 449943 Safeguard Custom Platform Script for Wikipedia.com

This script is for including Wikipedia as a new custom platform in SPP.

PrabhaQuest committed:
Feature 449943 Safeguard Custom Platform Script for Wikipedia.com

This feature includes Wikipedia as a new custom platform in SPP.

"SupportedScripts" Folder is added which contains CustomWikipedia.json and a ReadMe file in it.
